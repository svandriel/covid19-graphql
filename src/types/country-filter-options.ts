export interface CountryPredicateOptions {
    regions?: readonly string[] | null;
    excludeRegions?: readonly string[] | null;
    subRegions?: readonly string[] | null;
    excludeSubRegions?: readonly string[] | null;
    countries?: readonly string[] | null;
    excludeCountries?: readonly string[] | null;
}
