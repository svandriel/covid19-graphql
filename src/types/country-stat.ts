import { Moment } from 'moment';

export interface CountryStat {
    readonly countryCode: string;
    readonly lastUpdated: Moment;
    readonly confirmed: number;
    readonly deceased: number;
    readonly recovered: number;
}
