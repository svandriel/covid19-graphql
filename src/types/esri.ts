export interface EsriQuery {
    offset?: number;
    count?: number;
    outFields?: string[];
    orderBy?: Record<string, 'asc' | 'desc'>;
    where?: string;
}

export interface EsriResponse<T> {
    objectIdFieldName: string;
    uniqueIdField: {
        name: string;
        isSystemMaintained: boolean;
    };
    globalIdFieldName: string;
    geometryType?: string;
    spatialReference: {
        wkid: number;
        latestWkid: number;
    };
    fields: ReadonlyArray<{
        name: string;
        type: string;
        alias: string;
        sqlType: string;
        length?: number;
        domain: any;
        defaultValue: any;
    }>;
    exceededTransferLimit: boolean;
    features: ReadonlyArray<{
        attributes: T;
    }>;
}

export interface EsriCurrentStat {
    OBJECTID: number;
    Country_Region: string;
    Last_Update: number;
    Lat: number;
    Long_: number;
    Confirmed: number;
    Deaths: number;
    Recovered: number;
    Active: number;
    Incident_Rate: number | null;
    People_Tested: number | null;
}

export interface EsriHistoryStat {
    OBJECTID: number;
    Country_Region: string;
    Last_Update: number;
    Confirmed: number;
    Deaths: number;
    Recovered: number | null;
    Active: number | null;
    Delta_Confirmed: number;
    Delta_Recovered: number | null;
    Incident_Rate: number | null;
    People_Tested: number | null;
}
