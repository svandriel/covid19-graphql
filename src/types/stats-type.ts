export enum StatsType {
    Confirmed = 'CONFIRMED',
    Deceased = 'DECEASED',
    Recovered = 'RECOVERED',
}

export const statsTypes: StatsType[] = [StatsType.Confirmed, StatsType.Deceased, StatsType.Recovered];
