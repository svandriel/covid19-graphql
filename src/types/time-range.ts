import { Moment } from 'moment';

export interface TimeRange {
    from?: Moment | null;
    to?: Moment | null;
}
