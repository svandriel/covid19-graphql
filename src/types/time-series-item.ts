import { Moment } from 'moment';

import { ApiTimeSeriesItem } from '../generated/graphql-backend';

export interface TimeSeries {
    readonly countryCode: string;
    readonly state: string;
    items: readonly ApiTimeSeriesItem[];
}

export interface CountryStat {
    readonly countryCode: string;
    readonly lastUpdated: Moment;
    readonly confirmed: number;
    readonly deceased: number;
    readonly recovered: number;
}
