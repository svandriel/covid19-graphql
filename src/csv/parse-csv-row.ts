import moment from 'moment';

import { Country } from '../country-lookup';
import { ApiTimelineItem } from '../generated/graphql-backend';
import { StatsType } from '../types/stats-type';
import { Timeline } from '../types/timeline';
import { DATE_FORMAT_CSV } from '../util/date-formats';

export function parseCsvRow(
    lookupByName: Record<string, Country>,
    type: StatsType,
    row: Record<string, string>,
): Timeline {
    const countryName = row['Country/Region'];
    const state = row['Province/State'];
    const country = lookupByName[countryName];
    if (!country) {
        throw new Error(`Country not found: ${countryName}`);
    }
    return {
        countryCode: country?.code || 'NO CODE',
        state,
        items: parseTimeSeriesItems(type, row),
    };
}

export const invalidCountries = new Set<string>();

export function parseTimeSeriesItems(type: StatsType, row: Record<string, string>): ApiTimelineItem[] {
    return Object.keys(row)
        .filter(rowName => {
            return /^\d+\/\d+\/\d+$/.test(rowName);
        })
        .map(rowName => {
            const date = moment(rowName, DATE_FORMAT_CSV);
            if (date.isValid()) {
                const valueStr = row[rowName];
                const value = valueStr === '' ? 0 : parseInt(valueStr, 10);
                if (isNaN(value)) {
                    const countryName = row['Country/Region'];
                    const state = row['Province/State'];
                    throw new Error(
                        `Cannot parse '${valueStr}' into number for ${countryName}/${state} (type: ${type})`,
                    );
                }
                const x: ApiTimelineItem = {
                    date,
                    confirmed: type === StatsType.Confirmed ? value : 0,
                    recovered: type === StatsType.Recovered ? value : 0,
                    deceased: type === StatsType.Deceased ? value : 0,
                };
                return x;
            } else {
                throw new Error(`Cannot parse date: ${rowName}`);
            }
        });
}
