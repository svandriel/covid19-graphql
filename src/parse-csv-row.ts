import chalk from 'chalk';
import moment from 'moment';

import { countries } from './countries';
import { StatsType } from './types/stats-type';
import { TimeSeries, TimeSeriesItem } from './types/time-series';
import { DATE_FORMAT_CSV, DATE_FORMAT_REVERSE } from './util/date-formats';

export function parseCsvRow(type: StatsType, row: Record<string, string>): TimeSeries {
    const countryName = row['Country/Region'];
    const state = row['Province/State'];
    const countryCodeIso2 = countries[countryName];
    if (!countryCodeIso2) {
        console.error(`${chalk.red('ERROR')}: Invalid country name: ${chalk.cyan(countryName)}`);
    }
    return {
        provinceState: state,
        countryRegion: countryName,
        countryCodeIso2,
        lat: parseFloat(row.Lat),
        long: parseFloat(row.Long),
        items: parseTimeSeriesItems(type, row)
    };
}

export function parseTimeSeriesItems(type: StatsType, row: Record<string, string>): TimeSeriesItem[] {
    return Object.keys(row)
        .filter(rowName => {
            return /^\d+\/\d+\/\d+$/.test(rowName);
        })
        .map(rowName => {
            const date = moment(rowName, DATE_FORMAT_CSV);
            if (date.isValid()) {
                return {
                    date: date.format(DATE_FORMAT_REVERSE),
                    value: {
                        [StatsType.Confirmed]: 0,
                        [StatsType.Deceased]: 0,
                        [StatsType.Recovered]: 0,
                        [type]: parseInt(row[rowName], 10)
                    }
                };
            } else {
                throw new Error(`Cannot parse date: ${rowName}`);
            }
        });
}
