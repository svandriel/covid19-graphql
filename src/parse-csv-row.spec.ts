import { parseCsvRow } from './parse-csv-row';
import { StatsType } from './types/stats-type';

describe('Parse CSV row', () => {
    it('parses correctly', () => {
        expect(parseCsvRow(StatsType.Confirmed, makeData())).toMatchSnapshot();
    });
});

function makeData(): Record<string, string> {
    return {
        'Province/State': '',
        'Country/Region': 'Singapore',
        Lat: '1.2833',
        Long: '103.8333',
        '1/22/20': '0',
        '1/23/20': '1',
        '1/24/20': '3',
        '1/25/20': '3',
        '1/26/20': '4',
        '1/27/20': '5',
        '1/28/20': '7',
        '1/29/20': '7',
        '1/30/20': '10',
        '1/31/20': '13',
        '2/1/20': '16',
        '2/2/20': '18',
        '2/3/20': '18',
        '2/4/20': '24',
        '2/5/20': '28',
        '2/6/20': '28',
        '2/7/20': '30',
        '2/8/20': '33',
        '2/9/20': '40',
        '2/10/20': '45',
        '2/11/20': '47',
        '2/12/20': '50',
        '2/13/20': '58',
        '2/14/20': '67',
        '2/15/20': '72',
        '2/16/20': '75',
        '2/17/20': '77',
        '2/18/20': '81',
        '2/19/20': '84',
        '2/20/20': '84',
        '2/21/20': '85',
        '2/22/20': '85',
        '2/23/20': '89',
        '2/24/20': '89',
        '2/25/20': '91',
        '2/26/20': '93',
        '2/27/20': '93',
        '2/28/20': '93',
        '2/29/20': '102',
        '3/1/20': '106',
        '3/2/20': '108',
        '3/3/20': '110',
        '3/4/20': '110',
        '3/5/20': '117',
        '3/6/20': '130',
        '3/7/20': '138',
        '3/8/20': '150',
        '3/9/20': '150',
        '3/10/20': '160',
        '3/11/20': '178',
        '3/12/20': '178',
        '3/13/20': '200',
        '3/14/20': '212',
        '3/15/20': '226',
        '3/16/20': '243',
        '3/17/20': '266'
    };
}
