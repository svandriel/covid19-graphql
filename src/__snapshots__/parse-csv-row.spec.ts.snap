// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse CSV row parses correctly 1`] = `
Object {
  "countryCodeIso2": "SG",
  "countryRegion": "Singapore",
  "items": Array [
    Object {
      "date": "2020-01-22",
      "value": Object {
        "CONFIRMED": 0,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-01-23",
      "value": Object {
        "CONFIRMED": 1,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-01-24",
      "value": Object {
        "CONFIRMED": 3,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-01-25",
      "value": Object {
        "CONFIRMED": 3,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-01-26",
      "value": Object {
        "CONFIRMED": 4,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-01-27",
      "value": Object {
        "CONFIRMED": 5,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-01-28",
      "value": Object {
        "CONFIRMED": 7,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-01-29",
      "value": Object {
        "CONFIRMED": 7,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-01-30",
      "value": Object {
        "CONFIRMED": 10,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-01-31",
      "value": Object {
        "CONFIRMED": 13,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-01",
      "value": Object {
        "CONFIRMED": 16,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-02",
      "value": Object {
        "CONFIRMED": 18,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-03",
      "value": Object {
        "CONFIRMED": 18,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-04",
      "value": Object {
        "CONFIRMED": 24,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-05",
      "value": Object {
        "CONFIRMED": 28,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-06",
      "value": Object {
        "CONFIRMED": 28,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-07",
      "value": Object {
        "CONFIRMED": 30,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-08",
      "value": Object {
        "CONFIRMED": 33,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-09",
      "value": Object {
        "CONFIRMED": 40,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-10",
      "value": Object {
        "CONFIRMED": 45,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-11",
      "value": Object {
        "CONFIRMED": 47,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-12",
      "value": Object {
        "CONFIRMED": 50,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-13",
      "value": Object {
        "CONFIRMED": 58,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-14",
      "value": Object {
        "CONFIRMED": 67,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-15",
      "value": Object {
        "CONFIRMED": 72,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-16",
      "value": Object {
        "CONFIRMED": 75,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-17",
      "value": Object {
        "CONFIRMED": 77,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-18",
      "value": Object {
        "CONFIRMED": 81,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-19",
      "value": Object {
        "CONFIRMED": 84,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-20",
      "value": Object {
        "CONFIRMED": 84,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-21",
      "value": Object {
        "CONFIRMED": 85,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-22",
      "value": Object {
        "CONFIRMED": 85,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-23",
      "value": Object {
        "CONFIRMED": 89,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-24",
      "value": Object {
        "CONFIRMED": 89,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-25",
      "value": Object {
        "CONFIRMED": 91,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-26",
      "value": Object {
        "CONFIRMED": 93,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-27",
      "value": Object {
        "CONFIRMED": 93,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-28",
      "value": Object {
        "CONFIRMED": 93,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-02-29",
      "value": Object {
        "CONFIRMED": 102,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-01",
      "value": Object {
        "CONFIRMED": 106,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-02",
      "value": Object {
        "CONFIRMED": 108,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-03",
      "value": Object {
        "CONFIRMED": 110,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-04",
      "value": Object {
        "CONFIRMED": 110,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-05",
      "value": Object {
        "CONFIRMED": 117,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-06",
      "value": Object {
        "CONFIRMED": 130,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-07",
      "value": Object {
        "CONFIRMED": 138,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-08",
      "value": Object {
        "CONFIRMED": 150,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-09",
      "value": Object {
        "CONFIRMED": 150,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-10",
      "value": Object {
        "CONFIRMED": 160,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-11",
      "value": Object {
        "CONFIRMED": 178,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-12",
      "value": Object {
        "CONFIRMED": 178,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-13",
      "value": Object {
        "CONFIRMED": 200,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-14",
      "value": Object {
        "CONFIRMED": 212,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-15",
      "value": Object {
        "CONFIRMED": 226,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-16",
      "value": Object {
        "CONFIRMED": 243,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
    Object {
      "date": "2020-03-17",
      "value": Object {
        "CONFIRMED": 266,
        "DECEASED": 0,
        "RECOVERED": 0,
      },
    },
  ],
  "lat": 1.2833,
  "long": 103.8333,
  "provinceState": "",
}
`;
